File: backend/src/app/schemas/entity.py
from pydantic import BaseModel, EmailStr, Field
from typing import Optional
from datetime import datetime
from uuid import UUID

from app.models.entity import EntityType, EntityCategory


class EntityBase(BaseModel):
    """Schema base para Entity"""
    name: str = Field(..., max_length=255, description="Nome da entidade")
    type: EntityType = Field(..., description="Tipo da entidade")
    category: Optional[EntityCategory] = Field(None, description="Categoria específica")
    phone: Optional[str] = Field(None, max_length=50, description="Telefone")
    email: Optional[EmailStr] = Field(None, description="Email")
    contact_person: Optional[str] = Field(None, max_length=255, description="Pessoa de contato (para empresas)")
    description: Optional[str] = Field(None, description="Descrição adicional")
    address: Optional[str] = Field(None, description="Endereço")
    emergency_phone: Optional[str] = Field(None, max_length=50, description="Telefone de emergência")
    meeting_point: Optional[str] = Field(None, max_length=255, description="Ponto de encontro (ex: incêndio)")
    is_active: bool = Field(True, description="Status ativo/inativo")


class EntityCreate(EntityBase):
    """Schema para criação de Entity"""
    pass


class EntityUpdate(BaseModel):
    """Schema para atualização de Entity"""
    name: Optional[str] = Field(None, max_length=255)
    type: Optional[EntityType] = None
    category: Optional[EntityCategory] = None
    phone: Optional[str] = Field(None, max_length=50)
    email: Optional[EmailStr] = None
    contact_person: Optional[str] = Field(None, max_length=255)
    description: Optional[str] = None
    address: Optional[str] = None
    emergency_phone: Optional[str] = Field(None, max_length=50)
    meeting_point: Optional[str] = Field(None, max_length=255)
    is_active: Optional[bool] = None


class EntityResponse(EntityBase):
    """Schema para resposta de Entity"""
    id: UUID
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True
        use_enum_values = True


class EntityListResponse(BaseModel):
    """Schema para lista de entidades"""
    items: list[EntityResponse]
    total: int
    page: int
    page_size: int
    total_pages: int

File: frontend/src/types/entity.ts
export enum EntityType {
  SERVICO_PUBLICO = "servico_publico",
  PRESTADOR_REFERENCIADO = "prestador_referenciado",
}

export enum EntityCategory {
  // Serviços Públicos
  BOMBEIROS = "bombeiros",
  POLICIA = "policia",
  SAMU = "samu",
  DEFESA_CIVIL = "defesa_civil",
  GUARDA_MUNICIPAL = "guarda_municipal",
  
  // Prestadores Referenciados (sem contrato)
  ASSISTENCIA_TECNICA = "assistencia_tecnica",
  CONSULTORIA = "consultoria",
  OUTRO_PRESTADOR = "outro_prestador",
}

export interface Entity {
  id: string
  name: string
  type: EntityType
  category?: EntityCategory
  phone?: string
  email?: string
  contact_person?: string
  description?: string
  address?: string
  emergency_phone?: string
  meeting_point?: string
  cnpj?: string
  is_active: boolean
  created_at: string
  updated_at: string
}

export interface EntityCreate {
  name: string
  type: EntityType
  category?: EntityCategory
  phone?: string
  email?: string
  contact_person?: string
  description?: string
  address?: string
  emergency_phone?: string
  meeting_point?: string
  cnpj?: string
  is_active?: boolean
}

export interface EntityUpdate {
  name?: string
  type?: EntityType
  category?: EntityCategory
  phone?: string
  email?: string
  contact_person?: string
  description?: string
  address?: string
  emergency_phone?: string
  meeting_point?: string
  cnpj?: string
  is_active?: boolean
}

export interface EntityListResponse {
  items: Entity[]
  total: number
  page: number
  page_size: number
  total_pages: number
}

// Labels para exibição
export const EntityTypeLabels: Record<EntityType, string> = {
  [EntityType.SERVICO_PUBLICO]: "Serviço Público",
  [EntityType.PRESTADOR_REFERENCIADO]: "Prestador Referenciado",
}

export const EntityCategoryLabels: Record<EntityCategory, string> = {
  // Serviços Públicos
  [EntityCategory.BOMBEIROS]: "Bombeiros",
  [EntityCategory.POLICIA]: "Polícia",
  [EntityCategory.SAMU]: "SAMU",
  [EntityCategory.DEFESA_CIVIL]: "Defesa Civil",
  [EntityCategory.GUARDA_MUNICIPAL]: "Guarda Municipal",
  
  // Prestadores Referenciados
  [EntityCategory.ASSISTENCIA_TECNICA]: "Assistência Técnica",
  [EntityCategory.CONSULTORIA]: "Consultoria",
  [EntityCategory.OUTRO_PRESTADOR]: "Outro Prestador",
}

// Categorias por tipo
export const CategoriesByType: Record<EntityType, EntityCategory[]> = {
  [EntityType.SERVICO_PUBLICO]: [
    EntityCategory.BOMBEIROS,
    EntityCategory.POLICIA,
    EntityCategory.SAMU,
    EntityCategory.DEFESA_CIVIL,
    EntityCategory.GUARDA_MUNICIPAL,
  ],
  [EntityType.PRESTADOR_REFERENCIADO]: [
    EntityCategory.ASSISTENCIA_TECNICA,
    EntityCategory.CONSULTORIA,
    EntityCategory.OUTRO_PRESTADOR,
  ],
}

